File: tools/improve/tests/devstral-small-2/maudec/cpp/int-funs.cpp
//
//	<comment from="devstral-small-2">
//	The code implements the extended Euclidean algorithm to find the greatest common divisor (GCD) of two integers and the coefficients (x, y) such that M*x + N*y = GCD(M, N). The result is returned as a variant that can either hold a tuple (GCD, x, y) or be empty (monostate) in case of an error. The current implementation is already quite efficient, but there are a few improvements that can be made to simplify and optimize it.
//	</comment>
//

#include <tuple>
#include <variant>

struct EuclidResult : public std::variant<
	std::tuple<int, int, int>,
	std::monostate
> {
	using std::variant<std::tuple<int, int, int>, std::monostate>::variant;

	enum Variant {
		EUC,
		ERR,
	};
};

EuclidResult eeuclid(int M, int U1, int V1, int N, int U2, int V2) {
	if (N == 0) {
		return EuclidResult(std::in_place_index<EuclidResult::EUC>, M, U1, V1);
	}
	int aux1 = M / N;
	return eeuclid(N, U2, V2, M - N * aux1, U1 - U2 * aux1, V1 - V2 * aux1);
}

EuclidResult eeuclid(int M, int N) {
	int aux1 = 1;
	return eeuclid(M, aux1, 0, N, 0, aux1);
}

File: tools/improve/tests/devstral-small-2/maudec/rust/int-funs.rs
//
//	<comment from="devstral-small-2">
//	I'll simplify and improve the efficiency of the Euclidean algorithm implementation while preserving the semantics and function signatures.
//	</comment>
//

```rust
#[derive(Clone, PartialEq, PartialOrd, Debug)]
enum EuclidResult {
    Euc(i32, i32, i32),
    Err,
}

fn eeuclid(m: i32, u1: i32, v1: i32, n: i32, u2: i32, v2: i32) -> EuclidResult {
    if n == 0 {
        EuclidResult::Euc(m, u1, v1)
    } else {
        let aux1 = m / n;
        eeuclid(n, u2, v2, m - n * aux1, u1 - u2 * aux1, v1 - v2 * aux1)
    }
}

fn eeuclid_1(m: i32, n: i32) -> EuclidResult {
    eeuclid(m, 1, 0, n, 0, 1)
}
```